rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,392 @@
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we don't need ContainsDuplcates anymore because now we can
+	     // just check if duplicate count > 0!"
+	     DuplicateCount: 1 
+	}, nil
+}
+```
+
+Good
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    // TODO (avatus) DELETE IN 18
+    ContainsDuplicates bool
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we must keep ContainsDuplicates populated to support older clients!"
+	     ContainsDuplicates: true
+	     DuplicateCount: 1 
+	}, nil
+}
+``` 
+If using ConnectRPC (a proposed solution later in this RFD), then we must _not_ remove fields from requests/responses, even if unused. Mark as deprecated and move on.
+
+If a response shape must change entirely, prefer creating a new endpoint/RPC.
+
+### Updating Request body fields
+A Request body can receive a new field as long as the omission of the new field
+would not cause a break (i.e., an old client sending the old request without the
+new field). If the handler _requires_ this new field in order to function, then
+a new endpoint must be created with a version increment.
+
+
+### Adding new features to the web client
+If a feature requires a new endpoint, it must properly handle the case of a 404.
+This means making the "expected" 404 error clear to the user that an endpoint is
+not found due to a version mismatch (and preferably which minimum version is
+needed) rather than an ambiguous "Not Found" error.
+
+### New required fields from an API response in the web client
+If the updated feature _cannot_ function without receiving new fields from the
+API (for example, receiving a response from a proxy version N-1), refer to the
+API guidelines about creating a new endpoint. If the feature itself is degraded
+but still usable (be conservative with your discretion on this/ask product), a
+warning alert should be shown to the user with information about
+- What part of the feature may not work.
+- What version their proxy must be updated to in order to be fully operational. 
+
+An example of this in recent memory is Pinned Resources. The "pinned resources"
+tab was available in clients that _may_ have been communicating with proxies
+older than the required version. So a check was made to see if that proxy
+supported pinned resources, and if it wasn't, made the error known to the user
+that "Pinned Resources is supported by proxies version 14.1+" (or something like
+that).
+
+### Updating request fields in the web client
+Updating what fields are sent in a request body is ok as long as the API follows the updated request fields guidelines. 
```

Comment:
Yeah basically just the direction of the communication. I have them separated since I went with "server changes" and "web client changes". I think that you were able to assume the correct answer that it makes _enough sense_ but I will reword it to make it more clear regardless (or note or something). Thank you!

Reply:
Thank you for the feedback. I will update! 🙏

Reply:
I've removed this second paragraph and slightly changed the wording on the other to include both sides of the change.